parley/src/layout/line/greedy.rs

```diff
@@ -51,196 +117,248 @@ impl<'a, B: Brush> BreakLines<'a, B> {
-                            self.state.line.runs.end = self.state.i + 1;
-                            self.state.line.clusters.end = self.state.j + 1;
-                            self.state.line.x = next_x;
-                            self.state.j += 1;
-                            if commit_line(
-                                self.layout,
-                                &mut self.lines,
-                                &mut self.state.line,
-                                max_advance,
-                                alignment,
-                                BreakReason::Emergency,
-                                false,
-                            ) {
-                                self.state.runs = self.lines.runs.len();
-                                self.state.lines = self.lines.lines.len();
-                                self.state.line.x = 0.;
-                                let line = self.lines.lines.last().unwrap();
-                                self.state.prev_boundary = None;
-                                self.state.j += 1;
-                                return Some((line.metrics.advance, line.size()));
+                LayoutItemKind::TextRun => {
+                    let run_idx = item.index;
+                    let run_data = &self.layout.runs[run_idx];
+
+                    let run = Run::new(self.layout, run_data, None);
+                    let cluster_start = run_data.cluster_range.start;
+                    let cluster_end = run_data.cluster_range.end;
+
+                    println!("TextRun ({:?})", &run_data.text_range);
+
+                    // Iterate over remaining clusters in the Run
+                    while self.state.cluster_idx < cluster_end {
+                        let cluster = run.get(self.state.cluster_idx - cluster_start).unwrap();
+
+                        // Retrieve metadata about the cluster
+                        let is_ligature_continuation = cluster.is_ligature_continuation();
+                        let is_space = cluster.info().whitespace().is_space_or_nbsp();
+                        let boundary = cluster.info().boundary();
+
+                        let mut break_opportunity = false;
+
+                        // Handle boundary clusters
+                        match boundary {
+                            // A hard line break (e.g. CRLF or similar)
+                            Boundary::Mandatory => {
+                                println!("TextRun (Mandatory Break)");
+
+                                if !self.state.line.skip_mandatory_break {
+                                    self.state.prev_boundary = None;
+                                    self.state.line.items.end = self.state.item_idx + 1;
+                                    self.state.line.clusters.end = self.state.cluster_idx;
+
+                                    // We skip a mandatory break immediately after another mandatory break
```

Comment:
Is this the CRLF thing I think I heard/read somewhere (on Zulip?). What happens if we get LFLF, which I would expect to be two newlines.

Reply:
So long as it gets the right behaviour, that makes some sense. I must admit to still being a bit confused, but I trust that it all works out